## Importing the Dependencies.




In [2]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score
from sklearn.metrics import precision_score

In [3]:
df = pd.read_csv("C:\\Users\\abhis\\Downloads\\archive\\creditcard.csv")
print(df)


            Time         V1         V2        V3        V4        V5  \
0            0.0  -1.359807  -0.072781  2.536347  1.378155 -0.338321   
1            0.0   1.191857   0.266151  0.166480  0.448154  0.060018   
2            1.0  -1.358354  -1.340163  1.773209  0.379780 -0.503198   
3            1.0  -0.966272  -0.185226  1.792993 -0.863291 -0.010309   
4            2.0  -1.158233   0.877737  1.548718  0.403034 -0.407193   
...          ...        ...        ...       ...       ...       ...   
284802  172786.0 -11.881118  10.071785 -9.834783 -2.066656 -5.364473   
284803  172787.0  -0.732789  -0.055080  2.035030 -0.738589  0.868229   
284804  172788.0   1.919565  -0.301254 -3.249640 -0.557828  2.630515   
284805  172788.0  -0.240440   0.530483  0.702510  0.689799 -0.377961   
284806  172792.0  -0.533413  -0.189733  0.703337 -0.506271 -0.012546   

              V6        V7        V8        V9  ...       V21       V22  \
0       0.462388  0.239599  0.098698  0.363787  ... -0.01830

In [4]:
# Column Descriptions.

#1) Time(Float - Numerical) :- This column shows the time gap (in seconds) between each transaction and the first transaction in the dataset.It does not represent the actual date or clock time of the transaction.

#2) V1 to V28(Float - Numerical) :- These columns are numerical features created using PCA(Principal Component Analysis). The original transaction details were transformed to protect customer privacy.
#) Each column represents a hidden pattern or relationship in the data.

#3) Amount(Float - Numerical) :- This column represents the Transaction Amount in Money. Fraudulent Transactions often have Unusual or Abnormal Amounts. This Feature helps the model detect suspicious spending behavior.

#4) Class(Target Column - Integer (Binary / Categorical)) : This is the output column used for prediction. Value 0 means the transaction is normal (not fraud) and Value 1 means the Transaction is Fraudulent.

#) Therefore, Input Features: Time, V1–V28 and Amount and Output Feature: Class.

## EDA on this Dataset.

In [5]:
# Get the Quick Overview of the Dataset and Understand the structure of the Dataset.

df.shape  # Rows and Columns of the Dataset.
print("Infornmation about the Dataset")
df.info() # Complete Infornmation about the Columns of the Dataset.
df.columns # Infornmation i.e. Names of the Columns about the Dataset.
df.head(5) # Top 5 Rows of the Dataset.
df.tail(5) # Last 5 Rows of the Dataset.
df.dtypes # Getting Data - Types of each Column in the Dataset.
df.count()  # Count Non-Null Values in Each Column.
df.dtypes # Determining the Data - type of Columns in the Dataset.

Infornmation about the Dataset
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 284807 entries, 0 to 284806
Data columns (total 31 columns):
 #   Column  Non-Null Count   Dtype  
---  ------  --------------   -----  
 0   Time    284807 non-null  float64
 1   V1      284807 non-null  float64
 2   V2      284807 non-null  float64
 3   V3      284807 non-null  float64
 4   V4      284807 non-null  float64
 5   V5      284807 non-null  float64
 6   V6      284807 non-null  float64
 7   V7      284807 non-null  float64
 8   V8      284807 non-null  float64
 9   V9      284807 non-null  float64
 10  V10     284807 non-null  float64
 11  V11     284807 non-null  float64
 12  V12     284807 non-null  float64
 13  V13     284807 non-null  float64
 14  V14     284807 non-null  float64
 15  V15     284807 non-null  float64
 16  V16     284807 non-null  float64
 17  V17     284807 non-null  float64
 18  V18     284807 non-null  float64
 19  V19     284807 non-null  float64
 20  V20     284807 no

Time      float64
V1        float64
V2        float64
V3        float64
V4        float64
V5        float64
V6        float64
V7        float64
V8        float64
V9        float64
V10       float64
V11       float64
V12       float64
V13       float64
V14       float64
V15       float64
V16       float64
V17       float64
V18       float64
V19       float64
V20       float64
V21       float64
V22       float64
V23       float64
V24       float64
V25       float64
V26       float64
V27       float64
V28       float64
Amount    float64
Class       int64
dtype: object

In [6]:
# Checking for the Missing Vaues i.e. find out the Missing Values.

print(df.isnull()) # Null - Values in the form of Boolean Matrix(True - Missing Values , False - Not a mssing Values).
print(df.notnull()) # Opposite of isnull(True - Not a Missing Values , False - Missing Values)
print(df.isnull().sum()) # Count of the Missing Values of each Columns.
# By illustration, it is clear there are no missing values in any column in the Dataset.

         Time     V1     V2     V3     V4     V5     V6     V7     V8     V9  \
0       False  False  False  False  False  False  False  False  False  False   
1       False  False  False  False  False  False  False  False  False  False   
2       False  False  False  False  False  False  False  False  False  False   
3       False  False  False  False  False  False  False  False  False  False   
4       False  False  False  False  False  False  False  False  False  False   
...       ...    ...    ...    ...    ...    ...    ...    ...    ...    ...   
284802  False  False  False  False  False  False  False  False  False  False   
284803  False  False  False  False  False  False  False  False  False  False   
284804  False  False  False  False  False  False  False  False  False  False   
284805  False  False  False  False  False  False  False  False  False  False   
284806  False  False  False  False  False  False  False  False  False  False   

        ...    V21    V22    V23    V24

In [7]:
# Handling Duplicates in the Dataset.

print(df.duplicated())  # Reurns a Boolean series i.e. True denotes Duplicate Value and False denotes Not a Duplicate Value.
print(df.duplicated().sum()) # Sum of the Duplicate Values of the rows in the dataset(415 here).
df = df.drop_duplicates() #Dropping the Duplicate Rows in the Dataset.
print(df)

0         False
1         False
2         False
3         False
4         False
          ...  
284802    False
284803    False
284804    False
284805    False
284806    False
Length: 284807, dtype: bool
1081
            Time         V1         V2        V3        V4        V5  \
0            0.0  -1.359807  -0.072781  2.536347  1.378155 -0.338321   
1            0.0   1.191857   0.266151  0.166480  0.448154  0.060018   
2            1.0  -1.358354  -1.340163  1.773209  0.379780 -0.503198   
3            1.0  -0.966272  -0.185226  1.792993 -0.863291 -0.010309   
4            2.0  -1.158233   0.877737  1.548718  0.403034 -0.407193   
...          ...        ...        ...       ...       ...       ...   
284802  172786.0 -11.881118  10.071785 -9.834783 -2.066656 -5.364473   
284803  172787.0  -0.732789  -0.055080  2.035030 -0.738589  0.868229   
284804  172788.0   1.919565  -0.301254 -3.249640 -0.557828  2.630515   
284805  172788.0  -0.240440   0.530483  0.702510  0.689799 -0.377961   

## Univariate Analysis.

## Class Column

In [8]:
df['Class'].value_counts() # Counts how many normal (0) and fraud (1) transactions exist in the Dataset i.e. Frequency Analysis of Class Column.

# Observation -: Data is highly Imbalanced.

Class
0    283253
1       473
Name: count, dtype: int64

In [9]:
df['Class'].unique() # Shows all possible values in the Class Column.

array([0, 1], dtype=int64)

In [10]:
df['Class'].value_counts(normalize=True) * 100 # Percentage Distribution i.e. Converts counts into percentage.

# Observation : Normal transactions ≈ 99% and Fraud transactions ≈ 0.17% i.e. This means that the fraud cases are extremely rare.

Class
0    99.83329
1     0.16671
Name: proportion, dtype: float64

In [11]:
# separating the data for analysis
legit = df[df['Class'] == 0]
fraud = df[df['Class'] == 1]

In [12]:
print(legit.shape)
print(fraud.shape)

(283253, 31)
(473, 31)


In [13]:
# statistical measures of the data(legit data).
legit.Amount.describe()

count    283253.000000
mean         88.413575
std         250.379023
min           0.000000
25%           5.670000
50%          22.000000
75%          77.460000
max       25691.160000
Name: Amount, dtype: float64

In [14]:
# statistical measures of the data(fraud data).
fraud.Amount.describe() 

count     473.000000
mean      123.871860
std       260.211041
min         0.000000
25%         1.000000
50%         9.820000
75%       105.890000
max      2125.870000
Name: Amount, dtype: float64

In [15]:
# compare the values for both transactions as it Helps identify how fraud transactions differ from normal ones.
df.groupby('Class').mean()

,Time,V1,V2,V3,V4,V5,V6,V7,V8,V9,...,V20,V21,V22,V23,V24,V25,V26,V27,V28,Amount
Class,,,,,,,,,,,,,,,,,,,,,
0,94835.058093,0.013439,-0.009829,0.012853,-0.010440,0.006769,0.001251,0.010447,-0.002448,0.002613,...,-0.000489,-0.00115,-0.000160,0.000360,0.000393,-0.000301,0.000065,0.001409,0.000418,88.413575
1,80450.513742,-4.498280,3.405965,-6.729599,4.472591,-2.957197,-1.432518,-5.175912,0.953255,-2.522124,...,0.405043,0.46655,0.086639,-0.096464,-0.106643,0.040615,0.050456,0.213774,0.078270,123.871860


In [16]:
#) Observations:

#1)Time Column: 
# Transaction Type	Average Time
# Normal	            ~94,835
# Fraud	            ~80,450

# Fraud transactions usually happen earlier , their timing behavior is different and Time helps detect fraud.

#2) V1 – V28 Columns: These columns represent transaction behavior
#) They are mathematical features.
#) For Normal Transactions (Class = 0):
#) Values are very close to 0.
#) This means normal, expected behavior
#) For Fraud Transactions (Class = 1):
#) Values are far from 0.
#) Some are very negative, some very positive.
#) Fraud behavior looks very different.
#) The system can clearly see the difference
#) This is good for Machine learning.


#3) Amount Column:

# Transaction Type	Average Amount
# Normal	            ~88
# Fraud	            ~124

#) Fraud transactions usually involve more money
#) Fraudsters try to take larger amounts
#) Amount is a strong signal for fraud.



## Under-Sampling
- Reasons for choosing  Under-Sampling here:
1) Dataset Is Very Large : The dataset contains hundreds of thousands of normal transactions and Removing some normal transactions does not lose important information. So under-sampling is safe here.

2) Fraud Data Is Real and Limited : Fraud transactions are real, rare, and important and we should not duplicate or artificially create fraud
data. So, Under-sampling keeps only real data.

3) Avoid Overfitting(Compared to Oversampling) : Oversampling repeats fraud transactions, repeated data can cause the model to memorize, not learn, Under-Sampling avoids Duplication and So, therefore Logistic Regression stays Generalized.

4) Faster Training : Smaller dataset means faster model training, Less memory usage and Useful when computational resources are limited.

In [17]:
#1) For Under - Sampling, Building a Sample dataset containing similar distribution of Normal Transactions and Fraudulent Transactions.
#2) Number of Fraudulent Transactions --> 473.

In [18]:
fraud_count = fraud.shape[0]
fraud_count

473

In [19]:
legit_sample = legit.sample(n=473)

In [20]:
#) Concatenating two DataFrames.
new_dataset = pd.concat([legit_sample, fraud], axis=0)

In [23]:
new_dataset.head()

,Time,V1,V2,V3,V4,V5,V6,V7,V8,V9,...,V21,V22,V23,V24,V25,V26,V27,V28,Amount,Class
198276,132411.0,-0.175050,0.472561,1.203808,-0.523377,0.104453,-0.540269,0.374344,-0.056947,0.447687,...,0.444543,1.497426,-0.139123,0.050340,-0.702450,-0.279780,0.266626,0.238169,10.00,0
199100,132808.0,1.188351,-1.334561,-1.337815,1.368390,-0.110603,0.461728,0.270842,-0.010393,0.347166,...,0.318764,0.140389,-0.213994,0.297644,-0.108040,-0.725272,-0.043551,0.025640,408.61,0
168168,119096.0,-0.861530,0.121593,1.104623,-2.272322,0.442042,-0.572794,0.639236,-0.255672,-1.261141,...,0.090506,0.267924,-0.530579,-0.419507,0.736675,-0.202567,-0.278030,-0.144411,50.85,0
29577,35531.0,0.039727,0.324801,-0.492264,-1.779971,1.696449,3.822935,-1.683730,-1.691125,-0.224141,...,-1.529020,-0.670938,0.067151,1.012046,0.785662,0.750560,-0.112818,0.099721,16.88,0
47372,43190.0,0.968345,-0.453236,1.589424,1.172671,-0.962432,1.298112,-1.235473,0.678888,1.055388,...,0.294307,1.052262,-0.060461,-0.268408,0.240594,-0.157287,0.115667,0.026783,24.75,0


In [22]:
new_dataset.tail()

,Time,V1,V2,V3,V4,V5,V6,V7,V8,V9,...,V21,V22,V23,V24,V25,V26,V27,V28,Amount,Class
279863,169142.0,-1.927883,1.125653,-4.518331,1.749293,-1.566487,-2.010494,-0.882850,0.697211,-2.064945,...,0.778584,-0.319189,0.639419,-0.294885,0.537503,0.788395,0.292680,0.147968,390.00,1
280143,169347.0,1.378559,1.289381,-5.004247,1.411850,0.442581,-1.326536,-1.413170,0.248525,-1.127396,...,0.370612,0.028234,-0.145640,-0.081049,0.521875,0.739467,0.389152,0.186637,0.76,1
280149,169351.0,-0.676143,1.126366,-2.213700,0.468308,-1.120541,-0.003346,-2.234739,1.210158,-0.652250,...,0.751826,0.834108,0.190944,0.032070,-0.739695,0.471111,0.385107,0.194361,77.89,1
281144,169966.0,-3.113832,0.585864,-5.399730,1.817092,-0.840618,-2.943548,-2.208002,1.058733,-1.632333,...,0.583276,-0.269209,-0.456108,-0.183659,-0.328168,0.606116,0.884876,-0.253700,245.00,1
281674,170348.0,1.991976,0.158476,-2.583441,0.408670,1.151147,-0.096695,0.223050,-0.068384,0.577829,...,-0.164350,-0.295135,-0.072173,-0.450261,0.313267,-0.289617,0.002988,-0.015309,42.53,1


In [24]:
new_dataset['Class'].value_counts() 

Class
0    473
1    473
Name: count, dtype: int64

## Splitting the data into Features & Targets.

In [25]:
X = new_dataset.drop(columns='Class', axis=1)
Y = new_dataset['Class']

In [26]:
print(X)

            Time        V1        V2        V3        V4        V5        V6  \
198276  132411.0 -0.175050  0.472561  1.203808 -0.523377  0.104453 -0.540269   
199100  132808.0  1.188351 -1.334561 -1.337815  1.368390 -0.110603  0.461728   
168168  119096.0 -0.861530  0.121593  1.104623 -2.272322  0.442042 -0.572794   
29577    35531.0  0.039727  0.324801 -0.492264 -1.779971  1.696449  3.822935   
47372    43190.0  0.968345 -0.453236  1.589424  1.172671 -0.962432  1.298112   
...          ...       ...       ...       ...       ...       ...       ...   
279863  169142.0 -1.927883  1.125653 -4.518331  1.749293 -1.566487 -2.010494   
280143  169347.0  1.378559  1.289381 -5.004247  1.411850  0.442581 -1.326536   
280149  169351.0 -0.676143  1.126366 -2.213700  0.468308 -1.120541 -0.003346   
281144  169966.0 -3.113832  0.585864 -5.399730  1.817092 -0.840618 -2.943548   
281674  170348.0  1.991976  0.158476 -2.583441  0.408670  1.151147 -0.096695   

              V7        V8        V9  .

In [27]:
print(Y)

198276    0
199100    0
168168    0
29577     0
47372     0
         ..
279863    1
280143    1
280149    1
281144    1
281674    1
Name: Class, Length: 946, dtype: int64


## Split the data into Training data & Testing Data.

In [28]:
X_train, X_test, Y_train, Y_test = train_test_split(X, Y, test_size=0.2, stratify=Y, random_state=2)

In [29]:
print(X.shape, X_train.shape, X_test.shape)

(946, 30) (756, 30) (190, 30)


## Model Training(Logistic Regression Model).

In [30]:
model = LogisticRegression()

In [31]:
# Training the Logistic Regression Model with Training Data
model.fit(X_train, Y_train)

C:\Users\abhis\AppData\Local\Programs\Python\Python310\lib\site-packages\sklearn\linear_model\_logistic.py:465: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. OF ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(


LogisticRegression()

## Model Evaluation

## Accuracy Score

In [32]:
# Accuracy on Training Data.
X_train_prediction = model.predict(X_train)
training_data_accuracy = accuracy_score(X_train_prediction, Y_train)

In [33]:
print("Accuracy on Training data : ", training_data_accuracy)

Accuracy on Training data :  0.9378306878306878


1) The model was trained using Training Data.
2) Accuracy ≈ 93.9%
3) This means that the model correctly predicted about 94 out of 100 transactions in the training dataset.

In [34]:
# Accuracy on Test Data.
X_test_prediction = model.predict(X_test)
test_data_accuracy = accuracy_score(X_test_prediction, Y_test)

In [35]:
print("Accuracy score on Test Data : ", test_data_accuracy)

Accuracy score on Test Data :  0.9263157894736842


1) Test data was not seen by the model before.

2)  Accuracy ≈ 92.6%.
   
3)  This means that the model correctly predicted about 93 out of 100 unseen Transactions.

## Precision

In [36]:
#) Precision on Training Data 
X_train_prediction = model.predict(X_train)
training_data_precision = precision_score(Y_train, X_train_prediction)
print("Precision on Training data : ", training_data_precision)

Precision on Training data :  0.9610027855153204


- Training Precision(96.63%):-
- Out of 100 transactions marked as fraud(Predicted Fraud by the Model):
1) 97 are actually fraud
2) Only 3 are mistakes

- Therefore, The model learned very well during training.

In [37]:
# Precision on Test Data
X_test_prediction = model.predict(X_test)
test_data_precision = precision_score(Y_test, X_test_prediction)
print("Precision on Test data : ", test_data_precision)

Precision on Test data :  0.945054945054945


- Test Precision(94.51%):-
- Out of 100 transactions marked as fraud(Predicted Fraud by the Model):
1) 95 are actually fraud.
2) 5 are mistakes.
    
- Therefore, the model still performs very well in real life